### PR TITLE
[TRTLLM-14806][feat] Prefer Python V2 transceiver for LlamaForCausalLM and Gemma3ForCausalLM

### PR DESCRIPTION
**Pull request by moraxu**
## Dev Engineer Review

- `LlamaForCausalLM` and `Gemma3ForCausalLM` now prefer the Python V2 transceiver when `transceiver_runtime` is left at `auto` and the effective backend is NIXL.
- Explicit runtime selections are preserved, and UCX, MPI, or MOONCAKE still resolve to the C++ runtime.
- The scope is limited to checkpoints implemented by these two classes; `Llama4ForConditionalGeneration`, `MistralForCausalLM`, and `Gemma3ForConditionalGeneration` are unchanged.
- The upstream Llama/Gemma disaggregated tests, configs, and test-list registrations are retained.
- Llama-3.1 and Gemma-3 accuracy tests select NIXL while omitting the runtime, exercising automatic V2 selection end to end.
- Python V2 does not implement the serialized `DataTransceiverState` protocol and differs from C++ in experimental in-flight cancellation and synchronous timing output; those C++-specific tests remain on UCX/CPP, while the UCX logprobs RCCA path also remains unchanged.

## QA Engineer Review

- Unit coverage verifies that `auto` resolves to `PYTHON` for both model classes without overriding explicit runtime choices or selecting Python for unsupported backends.
- The retained Llama-3.1 matrix covers overlap scheduling, block reuse, NGram/EAGLE3 speculative decoding, guided decoding, TP/PP topologies, and multiple instances on the automatically selected V2 path.
- Gemma-3 coverage exercises automatic V2 selection with block reuse both enabled and disabled.
- Legacy UCX/CPP tests remain scheduled for the behaviors that are specific to the C++ runtime, and the final diff removes no upstream test-list or waive entries.

## Description

Prefer V2 transceiver backend for Gemma and Llama

## Test Coverage

```bash
pytest -v tests/unittest/llmapi/test_llm_args.py::TestTransceiverRuntimeAutoResolution
LLM_MODELS_ROOT=/path/to/models pytest -v tests/integration/defs/accuracy/test_disaggregated_serving.py::TestLlama3_1_8BInstruct
LLM_MODELS_ROOT=/path/to/models pytest -v tests/integration/defs/accuracy/test_disaggregated_serving.py::TestGemma3_1BInstruct
LLM_MODELS_ROOT=/path/to/models pytest -v 'tests/integration/defs/disaggregated/test_disaggregated.py::test_disaggregated_perf_metrics[TinyLlama-1.1B-Chat-v1.0]'
LLM_MODELS_ROOT=/path/to/models pytest -v 'tests/integration/defs/disaggregated/test_disaggregated.py::test_disaggregated_kv_cache_time_output[TinyLlama-1.1B-Chat-v1.0]'
LLM_MODELS_ROOT=/path/to/models pytest -v 'tests/integration/defs/disaggregated/test_disaggregated.py::test_disaggregated_logprobs_serving[llama-3.1-8b-instruct]'
LLM_MODELS_ROOT=/path/to/models pytest -v 'tests/integration/defs/disaggregated/test_disaggregated_single_gpu.py::test_arbitrary_kv_cache_transfer[False-TinyLlama-1.1B-Chat-v1.0]'
LLM_MODELS_ROOT=/path/to/models pytest -v 'tests/integration/defs/disaggregated/test_disaggregated_single_gpu.py::test_arbitrary_kv_cache_transfer_missing_blocks[False-TinyLlama-1.1B-Chat-v1.0]'
LLM_MODELS_ROOT=/path/to/models pytest -v 'tests/integration/defs/disaggregated/test_disaggregated.py::test_llama4_long_context_kv_cache_overflow[llama4-models/nvidia/Llama-4-Maverick-17B-128E-Instruct-FP8]'
```

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by Shixiaowei02** [comment]
Could you please check whether any existing tests need to be adjusted, similar to the changes made in [PR 16482](https://github.com/NVIDIA/TensorRT-LLM/pull/16482)?

**Comment by nv-xtf** [comment]
> Could you please check whether any existing tests need to be adjusted, similar to the changes made in [PR 16482](https://github.com/NVIDIA/TensorRT-LLM/pull/16482)?

I checked this, and the situation is slightly different from PR 16482. The existing Kimi test used DEFAULT, which caused the test harness to fall back to UCX/CPP, so it had to be explicitly changed to NIXL + PYTHON. Llama and Gemma already have corresponding NIXL + PYTHON disaggregated accuracy tests, so I don’t think similar test changes are needed here. 
The existing test_auto_dtype cases will continue to exercise UCX/CPP and can be retained as legacy coverage.

**Review by nv-xtf — APPROVED** [review]

**Review by Shixiaowei02 — APPROVED** [review]

**Comment by moraxu** [comment]
> Could you please check whether any existing tests need to be adjusted, similar to the changes made in [PR 16482](https://github.com/NVIDIA/TensorRT-LLM/pull/16482)?

I only updated one test using `TinyLlama` in `tests/integration/defs/disaggregated/test_disaggregated.py`

**Comment by moraxu** [comment]
I think my modified test errors here: `https://prod.blsm.nvidia.com/sw-tensorrt-top-1/job/LLM/job/main/job/L0_MergeRequest_PR/49570/testReport/junit/A10-PyTorch-3.disaggregated/test_disaggregated/test_disaggregated_perf_metrics_TinyLlama_1_1B_Chat_v1_0_/` - I will fix it shortly

**Review by ruodil — APPROVED** [review]

**Review by chienchunhung — APPROVED** [review]
Overall LGTM. 

One side-comment; not necessarily for this PR:

Consider defining `get_preferred_transceiver_runtime()` on the common model base class with a default `None`, then overriding it only for models that prefer `Python`. This gives all model classes a uniform interface and lets `_resolve_transceiver_runtime_auto()` call the hook directly instead of using `getattr`, while keeping the config/backend resolution policy outside the model layer.

**Comment by moraxu** [comment]
> Overall LGTM.
> 
> One side-comment; not necessarily for this PR:
> 
> Consider defining `get_preferred_transceiver_runtime()` on the common model base class with a default `None`, then overriding it only for models that prefer `Python`. This gives all model classes a uniform interface and lets `_resolve_transceiver_runtime_auto()` call the hook directly instead of using `getattr`, while keeping the config/backend resolution policy outside the model layer.

Yes, I can put up a separate PR for this, thanks for the suggestion

**Review by EmmaQiaoCh — APPROVED** [review]

**Review by zhaoyangwang-nvidia — APPROVED** [review]

**Review by 2ez4bz — APPROVED** [review]

**Review by yingguo-trt — APPROVED** [review]

**Comment by BowenFu** [comment]
Holding on this one. The 14 lines of product change are a genuine default flip, and the description does not match what the head actually does.

**The flip.** `transceiver_runtime` defaults to `"auto"` (`llm_args.py:3755`), `_resolve_transceiver_runtime_auto` adopts the model preference when the effective backend is NIXL (`llm_utils.py:601-623`), and `_resolve_default_backend` returns `"NIXL"` for `backend="DEFAULT"` with no legacy env var set (`llm_args.py:3802-3807`). So every Llama and Gemma3 disaggregated deployment on a default `cache_transceiver_config` moves from the C++ transceiver to the Python one. That is the two most widely deployed model families on the KV-transfer path, and `LlamaForCausalLM` also covers TinyLlama and the whole Llama-3.x line. Worth stating explicitly in the description — right now `## Description` is one line and `## Test Coverage` is empty.

**The description is stale in a way that matters.** It says:

> `test_disaggregated_perf_metrics` and `test_disaggregated_kv_cache_time_output` now force **UCX over NIXL** via `TRTLLM_USE_NIXL_KVCACHE=0` and `TRTLLM_USE_UCX_KVCACHE=1`

At head `92f648430` both functions are **deleted**, not adjusted — `grep -c "def test_disaggregated_perf_metrics"` on the head blob returns 0, same for `test_disaggregated_kv_cache_time_output`, along with `disagg_config_metrics.yaml` and both l0_a10 entries. The auto-generated QA section on the same description already says "Coverage in `tests/integration/test_lists/`: Not verified — Verdict: needs follow-up". With eight approvals on the PR, it is worth being sure they were reading the head and not that paragraph.

**Three coverage deletions I do not think the flip justifies:**

1. `test_llama4_long_context_kv_cache_overflow` and `disagg_config_llama4_kv_cache_overflow.yaml` — an RCCA test for nvbug 5555681 (KV cache buffer overflow, long context, 8×GPU Maverick FP8). `Llama4ForConditionalGeneration` (`modeling_llama.py:1489`) is a separate class and does **not** get the new override, so this test covers a path this PR does not change. Why is it going away?

2. `disagg_config_ctxtp2_gentp2_llama31_8b_ucx.yaml` is deleted and `test_disaggregated_logprobs_serving` (RCCA for NVBug 5926823) is repointed to NIXL+Python. But per `llm_utils.py:616-621`, when the backend is UCX/MPI/MOONCAKE the preference falls back to the C++ transceiver — Llama-on-CPP is still a live, reachable default. This deletes the only multi-GPU-TP coverage of a path that remains in service.

3. The `TestLlama3_1_8BInstruct` matrix — `test_auto_dtype`'s 16 overlap-scheduler × block-reuse combinations, `test_eagle3`, `test_guided_decoding`, `test_guided_decoding_with_eagle3`, `test_multi_instance`, `test_gen_only_spec_dec` — is removed from source, from `l0_dgx_h100.yml`/`l0_dgx_h200.yml`, and ~40 entries from `llm_function_core.txt`. (13 of the removed entries were already waived, so I am only counting the ones that were actually running.) That leaves `test_kv_cache_v2_nixl_python` as essentially the entire pre-merge signal for Llama disaggregated — on the very release where its default runtime changes. Those parametrizations are about block reuse and overlap scheduling, which are not CPP-specific; pinning them with `transceiver_runtime="CPP"` would keep them meaningful.

Point 3 is the crux: the flip and the removal of the coverage that would catch a regression in it are landing in the same PR, so the new default arrives with less pre-merge signal than the old one had.

For what it is worth, the two `transceiver_runtime="CPP"` pins added to `test_arbitrary_kv_cache_transfer*` in `test_disaggregated_single_gpu.py` are exactly right, and the comment explaining them is good. That is the same treatment I would want for the tests above rather than deletion.

**Comment by BowenFu** [comment]
Follow-up to the above — an independent pass turned up two things I had not checked, plus a number worth having:

**Capability gaps under the new default.** The Python transceiver does not implement the C++ serialized `DataTransceiverState` protocol used for arbitrary / reuse-tree transfers — which is exactly why the two `test_arbitrary_kv_cache_transfer*` cases had to be pinned to `"CPP"` in this diff. Experimental in-flight cancellation and the synchronous timing-output behaviour also differ between the two runtimes. Those are user-visible for Llama and Gemma deployments that are on the C++ path today and will be moved off it without changing their config. Worth stating in the description which C++ capabilities are and are not present on the Python side for these models.

**The coverage number.** Counting only entries that were actually running (i.e. excluding the 13 waived at `waives.txt:1-13`): `l0_dgx_h100.yml` had 30 Llama-3.1 disaggregated entries and `l0_dgx_h200.yml` had 6, so 23 running. After this PR H100 keeps 2 and H200 keeps none — **23 → 2**.

And the part I would most like your read on: both survivors set `"transceiver_runtime": "PYTHON"` explicitly. The only test of the new behaviour is `test_llama_and_gemma_model_preferences_adopted`, which calls `_resolve_transceiver_runtime_auto` directly and asserts `"auto" → "PYTHON"`. So **no test exercises the new automatic resolution end to end** — nothing loads a model with a shipped DEFAULT config and confirms the resulting run is correct on the runtime it now silently selects. That is the gap I would want closed before this merges, and it is cheap: re-add one of the deleted `test_auto_dtype` cases with the transceiver config left at its default.

File: tests/integration/defs/disaggregated/test_disaggregated.py [hunk 1]
```diff
@@ -360,8 +354,8 @@ def get_test_config(test_desc, example_dir, test_root):
         f"{test_configs_root}/disagg_config_cancel_stress_test.yaml",
         "cancel_stress_test_large":
         f"{test_configs_root}/disagg_config_cancel_stress_test_large.yaml",
-        "llama31_8b_ucx":
-        f"{test_configs_root}/disagg_config_ctxtp2_gentp2_llama31_8b_ucx.yaml",
+        "llama31_8b_nixl_python":
```
Comment (brnguyen2) [inline tests/integration/defs/disaggregated/test_disaggregated.py]: `test_disaggregated_logprobs_serving` is the NVBug 5926823 regression test; repointing it from the UCX config to NIXL+PYTHON means the original repro transport is no longer exercised. Consider parametrizing over both configs, or at least keeping the UCX one so the RCCA path stays covered.

File: tests/integration/defs/disaggregated/test_configs/disagg_config_ctxtp2_gentp2_llama31_8b_nixl_python.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+backend: pytorch
+disable_overlap_scheduler: true
+gather_generation_logits: true
+context_servers:
+  num_instances: 1
+  tensor_parallel_size: 2
+  pipeline_parallel_size: 1
+  cache_transceiver_config:
+    backend: NIXL
+    transceiver_runtime: PYTHON
```
Comment (brnguyen2) [inline tests/integration/defs/disaggregated/test_configs/disagg_config_ctxtp2_gentp2_llama31_8b_nixl_python.yaml]: Pinning `transceiver_runtime: PYTHON` explicitly means this config never exercises the thing this PR actually changes. Leaving it out (or `backend: DEFAULT`) would resolve to NIXL + PYTHON via the new model preference, giving an end-to-end check that the auto-resolution fires for Llama — which is currently only covered by a unit test on `_resolve_transceiver_runtime_auto`.

File: tensorrt_llm/_torch/models/modeling_llama.py [hunk 3]
```diff
@@ -1134,6 +1134,13 @@ def forward(
 @register_auto_model("LlamaForCausalLM")
 class LlamaForCausalLM(SpecDecOneEngineForCausalLM[LlamaModel, LlamaConfig]):
 
+    @classmethod
+    def get_preferred_transceiver_runtime(
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/models/modeling_llama.py:1138]: Worth stating in the PR description: this only affects checkpoints whose `architectures[0]` is `LlamaForCausalLM`. `Llama4ForConditionalGeneration` and `MistralForCausalLM` are separate classes (lines 1505/1617) and keep the C++ path, as does `Gemma3ForConditionalGeneration` on the Gemma side — the title reads as if all Llama/Gemma variants flip. This is also why removing `test_llama4_long_context_kv_cache_overflow` isn't justified by this change.

**Comment by moraxu** [comment]
Thanks, addressed the points as follows:
1. **Default flip and scope:** The description now states that automatic V2 selection applies only to `LlamaForCausalLM` and `Gemma3ForCausalLM`. Llama 4, Mistral, and multimodal Gemma 3 classes remain unchanged.
2. **Stale description:** The Dev/QA sections now reflect the current six-file diff instead of the earlier deletion-heavy revision.
3. **Llama 4 RCCA:** `test_llama4_long_context_kv_cache_overflow` and its configuration were restored unchanged.
4. **UCX logprobs RCCA:** The original UCX configuration and `test_disaggregated_logprobs_serving` path were restored, preserving multi-GPU C++ coverage.
5. **Llama-3.1 correctness matrix:** The deleted accuracy tests and their H100/H200/QA registrations were restored. They now select literal NIXL while omitting `transceiver_runtime`, exercising automatic Python V2 selection end to end.
6. **C++-specific behavior:** Timing metrics and CSV-output tests now explicitly force UCX/C++, while arbitrary reuse-tree transfers explicitly set `transceiver_runtime="CPP"`.
7. **Capability differences:** The description now calls out serialized `DataTransceiverState`, experimental cancellation, and synchronous timing-output differences between V1 and V2.
8. **Coverage count:** The previous 23 active Llama-3.1 correctness nodes remain scheduled instead of dropping to two.
9. **Automatic-resolution coverage:** In addition to the unit test, the restored Llama and Gemma accuracy tests now validate automatic runtime selection through real disaggregated serving.
